README.md
# Project python bus

The pybus.py script parse the St-Jean-Sur-Le-Richelieu autobus timeline.

## api call

    * /          - return all bus trajectories ( sjsr and mtrl) and max 10
    * ?dest=sjsr - return only sjsr bus trajectories (combines with max=x)
    * ?dest=mtrl - return only mtrl bus trajectories (combines with max=x)
    * ?max=x     - return only x bus trajectories (combines with ?dest=value)

## Requirements

Install the required python packages

```bash
pip install --no-cache-dir -r requirements.txt
```

## Running the tests

```bash
python pybus_test.py -v
```

## Authors

* **Jeremie Bellavance** - *Maintainer*


## Acknowledgments

* http://python.org/
* https://www.flaskapi.org/
* https://serverless.com/
* https://aws.amazon.com/

CHANGELOG
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
### [0.1.3] - 2020-01-20
### fixed sorting bug
- ajusted minutes sorting
- changed the default max from 5 to 10
### [0.1.2] - 2019-12-31
### custom dns api.mayoche.info + sls deploy
- modified serverless to add custom dns
- modified gitignore file to remove node modules
- remove ignored files
- fixed prod deploymenent to AWS with sls framework
### [0.1.1] - 2019-10-29 - 2019-11-15
### Unicode character now supported + DynamoDB integration
- FIXED: Wrong speed bus bug caused by this symbol -> ☀
- changed the serverless python version from 3.6 to 3.7
- Other fix to timezone because AWS lambda it UTC
- fixed serverless deployment dashboard
- get content from dynamoDB
- update the dynamoDB if empty
- added dynamoDB table in the serverless configuration
- increased the lambda cold start performance significantly
- modified the test pipeline to test with dynamodb prod table
### [0.1.0] - 2019-10-26
### Timezone fix + serverless integration
- Timezone fixed to America/Montreal
- added serverless integration for AWS deployments
### [0.0.9] - 2019-08-21
### minor improvements
- minor fixes
- syntax review
### [0.0.8] - 2019-07-03
### bug fix time
- Bug fix time, using datetime
### [0.0.7] - 2019-06-26
### cache feature added
- The api now caches the city site response
- The api now tries 10 times to get city website data
### [0.0.6] - 2019-06-26
### max feature added
- The api now return x futures buses from ?max=x variable
- function populate_complete create to remove repeated code
## [0.0.5] - 2019-06-24
### time validation added
- The api now return only 10 futures buses from current time GMT
## [0.0.4] - 2019-06-20
### specific destination method added
- The api can now return only a specific location (/?dest=)
- Improvement of switch speed validation
- timeout of bus site augmented to 30 sec
## [0.0.3] - 2019-06-13
### deployment automation
- Production automation added
## [0.0.2] - 2019-06-13
### API, tests
- Converted into flask API
- Testing suite added
## [0.0.1] - 2019-06-12
### First version
- Fetches all the data from the city website

pybus_test.py
#!/usr/bin/env python
import unittest
from pybus import app


class Testcalltoapi(unittest.TestCase):

    def setUp(self):
        self.app = app.test_client()
        self.app.testing = True

    def test_noargs(self):
        rv = self.app.get('/')
        self.assertEqual(rv.status_code, 200)

    def test_sjsr(self):
        rv = self.app.get('/?dest=sjsr')
        self.assertEqual(rv.status_code, 200)

    def test_mtrl(self):
        rv = self.app.get('/?dest=mtrl')
        self.assertEqual(rv.status_code, 200)

    def test_all(self):
        rv = self.app.get('/?dest=all')
        self.assertEqual(rv.status_code, 200)


if __name__ == '__main__':
    unittest.main()


pybus.py
#!/usr/bin/python
import re
import urllib.request
from datetime import datetime
import os
import boto3
from botocore.exceptions import ClientError
from pytz import timezone
import pytz
from bs4 import BeautifulSoup
from flask import request
from flask_api import FlaskAPI, status

app = FlaskAPI(__name__)

# dynamodb table information
USERS_TABLE = os.environ['USERS_TABLE']
dynamodb = boto3.resource('dynamodb', 'us-east-1')
table = dynamodb.Table(USERS_TABLE)


@app.before_first_request
###
# try to get information from dynamodb
###
def get_or_update_dynamodb_data():
    ###
    # TODO
    # fix AWS credentials for the git server
    ###
    try:
        response = table.scan()
        global dynamodb_data
        dynamodb_data = response['Items']
    except ClientError as e:
        print(e.response['Error']['Message'])
        print("Can't fetch data from dynamodb")
    if len(dynamodb_data) <= 4:
        ###
        # Updating the dynamoDB only if empty
        ###
        url = ('http://www.ville.saint-jean-sur-richelieu.qc.ca/' +
               'transport-en-commun/Documents/horaires/96.html')
        for tries in range(1, 11):
            try:
                response_web = urllib.request.urlopen(url, timeout=30)
                html_doc = response_web.read()
            except Exception as E:
                print("Try " + str(tries) + "/10 failed for " + url + " exception is : " + str(E))
                continue
        ###
        # Parsing the collected data
        ###
        soup = BeautifulSoup(html_doc, 'html.parser')
        dir_list = soup.find_all('div', attrs={"id": "div-horaires"})
        dir_to_mtrl_table = dir_list[0].find('table')
        dir_from_mtrl_table = dir_list[3].find('table')

        speed_to_mtrl = dir_to_mtrl_table.find_all('div', attrs={"align": "center"})
        start_to_mtrl = dir_to_mtrl_table.find_all('tr')[1]
        end_to_mtrl = dir_to_mtrl_table.find_all('tr')[-1]

        speed_to_sjsr = dir_from_mtrl_table.find_all('div', attrs={"align": "center"})
        start_to_sjsr = dir_from_mtrl_table.find_all('tr')[1]
        end_to_sjsr = dir_from_mtrl_table.find_all('tr')[-1]

        start_to_mtrl_lst = []
        end_to_mtrl_lst = []
        start_to_sjsr_lst = []
        end_to_sjsr_lst = []

        def populate_list_direction(city_start, list_start, city_end, list_end):
            for i in city_start:
                i = str(i)
                if i != '\n':
                    list_start.append(re.sub("<.*?>", "", i))
            for i in city_end:
                i = str(i)
                if i != '\n':
                    list_end.append(re.sub("<.*?>", "", i))

        populate_list_direction(start_to_mtrl, start_to_mtrl_lst, end_to_mtrl, end_to_mtrl_lst)
        populate_list_direction(start_to_sjsr, start_to_sjsr_lst, end_to_sjsr, end_to_sjsr_lst)
        ###
        # Adding Saint-Jean-Sur-Le-Richelieu bus runs information
        ###
        for start, end, speed in zip(start_to_sjsr_lst, end_to_sjsr_lst, speed_to_sjsr):
            try:
                speed = str(speed)
                destination = 'sjsr'
                concat_response = (destination + ';'
                                   + start + ';'
                                   + end + ';'
                                   + speed)
                insert_data = table.put_item(
                    TableName=USERS_TABLE,
                    Item={
                        'data': str(concat_response)
                    }
                )
            except ClientError as e:
                print(e.response['Error']['Message'])
        ###
        # Adding Montreal bus runs information
        ###
        for start, end, speed in zip(start_to_mtrl_lst, end_to_mtrl_lst, speed_to_mtrl):
            try:
                speed = str(speed)
                destination = 'mtrl'
                concat_response = (destination + ';'
                                   + start + ';'
                                   + end + ';'
                                   + speed)
                insert_data = table.put_item(
                    TableName=USERS_TABLE,
                    Item={
                        'data': str(concat_response)
                    }
                )
            except ClientError as e:
                print(e.response['Error']['Message'])


@app.route("/", methods=['GET'])
def parse_bus():
    if request.method == 'GET':
        ###
        # Defining Montreal timezone to match data source timezone
        ###
        localtime = datetime.now(pytz.utc)
        montreal_tz = timezone('America/Montreal')
        date = localtime.astimezone(montreal_tz)
        date_time_hours = date.strftime("%H")
        date_time_minutes = date.strftime("%M")
        ###
        # Parsing given parameters
        ###
        if request.args.get("max"):
            direction_max = request.args.get("max", "")
            direction_max = int(direction_max)
        else:
            direction_max = 10
        if request.args.get("dest"):
            direction = request.args.get("dest", "")
        else:
            direction = 'all'
        destination = ['sjsr', 'mtrl', 'all']
        if direction.lower() not in destination:
            return_message = "Variable dest=" + direction.lower() + " invalid. Must be dest=" + destination[
                0].lower() \
                             + " or dest=" + destination[1].lower()
            return return_message, status.HTTP_400_BAD_REQUEST

        ###
        # Filtering output value to give
        ##
        def list_of_speeds(argument):
            switcher = {
                '<div align="center">S</div>': "Super Express  ",
                '<div align="center">S ☀</div>': "Super Express ☀",
                '<div align="center">E</div>': "Express        ",
                '<div align="center">E ☀</div>': "Express ☀     ",
                '<div align="center">L</div>': "Locale         ",
                '<div align="center">L ☀</div>': "Locale ☀      ",
                '<div align="center">A</div>': "Autoroute 30   ",
                '<div align="center">A ☀</div>': "Autoroute 30 ☀ ",
            }
            return switcher.get(argument, "Wrong speed    ")

        complete_return_value_mtrl = []
        complete_return_value_sjsr = []
        complete_return_value = []

        def populate_complete(bus_data):
            for entry in bus_data:
                # Entry example
                # sjsr;17:06;17:46;<div align="center">S</div>
                dest, start, end, speed = entry['data'].split(';')
                loop_hours, loop_minutes = start.split(':')
                combined_loop_minutes = int(loop_hours)*60 + int(loop_minutes)
                combined_date_minutes = int(date_time_hours)*60 + int(date_time_minutes)
                if combined_loop_minutes >= combined_date_minutes:
                    value = ("Autobus destination " + dest + " : "
                             + list_of_speeds(speed) + " Depart:"
                             + start + " Arriver:" + end)
                    if dest == 'sjsr':
                        complete_return_value_sjsr.append(value)
                    elif dest == 'mtrl':
                        complete_return_value_mtrl.append(value)

        populate_complete(dynamodb_data)
        ###
        # Sorting bus rides based on the start hour
        ###

        def custom_sort(t):
            dest, speed, hour_start, minutes_start, hour_end, minutes_end = t.split(':')
            return hour_start, minutes_start
        complete_return_value_sjsr = sorted(complete_return_value_sjsr, key=custom_sort)
        complete_return_value_mtrl = sorted(complete_return_value_mtrl, key=custom_sort)
        ###
        # Adding only the bus rides up to direction_max
        ###
        if direction == 'sjsr':
            complete_return_value = complete_return_value_sjsr[:direction_max]
        elif direction == 'mtrl':
            complete_return_value = complete_return_value_mtrl[:direction_max]
        else:
            complete_return_value.extend(complete_return_value_sjsr[:direction_max])
            complete_return_value.extend(complete_return_value_mtrl[:direction_max])

        return complete_return_value, status.HTTP_200_OK


if __name__ == "__main__":
    app.run()
